### PR TITLE
gh-64490: varargs support for argument clinic and refactor print with AC

### PR DESCRIPTION
**Pull request by isidentical**
- [x] `f(*args, /)`
- [x] `f(x, y, *args, /)`
- [x] `f(*args)`
- [x] `f(x, y, *args)`
- [x] `f(x, /, y, *args)`
- [x] `f(x, *args, y=3)`
- [x] `f(*args, y=3, z=5)`

https://bugs.python.org/issue20291

* Issue: gh-64490

### CODE REVIEW

File: Misc/NEWS.d/next/Tools-Demos/2020-02-25-18-22-09.bpo-20291.AyrDiZ.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Vararg (`*args`) notation support for Argument Clinic. Patch by Batuhan
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Tools-Demos/2020-02-25-18-22-09.bpo-20291.AyrDiZ.rst]:
```suggestion
Added support for :term:`var-positional` parameter in Argument Clinic. Patch by Batuhan
```

File: Python/bltinmodule.c [hunk 2]
```diff
@@ -1821,23 +1821,31 @@ builtin_pow_impl(PyObject *module, PyObject *base, PyObject *exp,
     return PyNumber_Power(base, exp, mod);
 }
 
+/*[clinic input]
+print as builtin_print
+
+    *args: object
+    sep: object(c_default="NULL") = ' '
+    end: object(c_default="NULL") = '\\n'
```
Comment (serhiy-storchaka) [inline Python/bltinmodule.c]:
```suggestion
    end: object(c_default="NULL") = '\n'
```
or
```suggestion
    end: object(c_default="Py_None") = '\n'
```
or
```suggestion
    end: object = None
```

File: Python/bltinmodule.c [hunk 3]
```diff
@@ -1821,23 +1821,31 @@ builtin_pow_impl(PyObject *module, PyObject *base, PyObject *exp,
...
+    file: object(c_default="NULL") = None
```
Comment (serhiy-storchaka) [inline Python/bltinmodule.c]:
```suggestion
    file: object = None
```
And remove the special case for NULL below.

File: Python/bltinmodule.c [hunk 4]
```diff
@@ -1821,23 +1821,31 @@ builtin_pow_impl(PyObject *module, PyObject *base, PyObject *exp,
...
+    flush: bool(c_default=0) = False
```
Comment (serhiy-storchaka) [inline Python/bltinmodule.c]:
```suggestion
    flush: bool = False
```

File: Python/bltinmodule.c [hunk 5]
```diff
@@ -1821,23 +1821,31 @@ builtin_pow_impl(PyObject *module, PyObject *base, PyObject *exp,
...
+
+Prints the values to a stream, or to sys.stdout by default.
+
+Optional keyword arguments:
+file:  a file-like object (stream); defaults to the current sys.stdout.
```
Comment (serhiy-storchaka) [inline Python/bltinmodule.c]: It is better to add descriptions directly to parameters.
```
    file: object = None
        a file-like object (stream); defaults to the current sys.stdout.
```

File: Python/bltinmodule.c [hunk 6]
```diff
@@ -1821,23 +1821,31 @@ builtin_pow_impl(PyObject *module, PyObject *base, PyObject *exp,
     return PyNumber_Power(base, exp, mod);
 }
 
+/*[clinic input]
+print as builtin_print
+
+    *args: object
+    sep: object(c_default="NULL") = ' '
```
Comment (serhiy-storchaka) [inline Python/bltinmodule.c]: May be also
```suggestion
    sep: object(c_default="Py_None") = ' '
```
or
```suggestion
    sep: object = None
```

File: Include/modsupport.h [hunk 7]
```diff
@@ -50,6 +50,8 @@ PyAPI_FUNC(PyObject *) Py_BuildValue(const char *, ...);
 PyAPI_FUNC(PyObject *) _Py_BuildValue_SizeT(const char *, ...);
 
 
+#define NO_VARARG -1
```
Comment (serhiy-storchaka) [inline Include/modsupport.h]: Why not use `PY_SSIZE_T_MAX`?

**Review by serhiy-storchaka — COMMENTED** [review]
Interesting PR. Thank you that taken this issue Batuhan! :+1:  But it will take a time to make a review.

File: Include/modsupport.h (hunk 7 above)
Comment (isidentical) [reply Include/modsupport.h]: Sure.

File: Python/bltinmodule.c (hunk 5 above)
Comment (isidentical) [reply Python/bltinmodule.c]: Yes, thanks.

**Comment by isidentical** [comment]
Thank you for your reviews @serhiy-storchaka. @pablogsal and I were working on this as a part of mentorship. I initially implemented `_PyArgno_UnpackKeywordsWithVararg` as the main function and called it inside of `_PyArg_UnpackKeywords` but then we noticed a performance drop on benchmarks. So I splitted the function. If I'm not mistaken the current version has no significant performance affects.

**Comment by pablogsal** [comment]
@serhiy-storchaka Could you review this PR if you have the time? If you are busy I can finish the review myself and try to land it so it reaches 3.11 :)

**Comment by pablogsal** [comment]
@isidentical Can you rebase this and we can land it?

**Review by pablogsal — APPROVED** [review]

**Comment by pablogsal** [comment]
@isidentical Land if the buildbots are happy. Don't forget to also change the docs for the argument clinic in the devguide with any updates!

**Comment by gpshead** [comment]
The Doc/howto/clinic.rst docs still need updating.

**Pull request merged**